edb/edgeql/compiler/func.py

```diff
@@ -356,14 +460,54 @@ def compile_FunctionCall(
         for arg in res.args.values():
             pathctx.register_set_in_scope(
                 arg.expr,
-                optional=arg.param_typemod == ft.TypeModifier.OptionalType,
+                optional=(
+                    arg.param_typemod == ft.TypeModifier.OptionalType
+                    or arg.is_default
+                ),
                 ctx=ctx,
             )
 
     ir_set = setgen.ensure_set(res, typehint=rtype, path_id=path_id, ctx=ctx)
     return stmt.maybe_add_view(ir_set, ctx=ctx)
 
 
+class ArgumentInliner(ast.NodeTransformer):
+
+    mapped_args: dict[irast.PathId, irast.PathId]
+
+    def __init__(
+        self,
+        inline_args: dict[int | str, irast.CallArg | irast.Set],
+        ctx: context.ContextLevel,
+    ) -> None:
+        super().__init__()
+        self.inline_args = inline_args
+        self.ctx = ctx
+        self.mapped_args = {}
+
+    def visit_Set(self, node: irast.Set) -> irast.Base:
+        if (
+            isinstance(node.expr, irast.Parameter)
+            and node.expr.name in self.inline_args
+        ):
+            arg = self.inline_args[node.expr.name]
+            if isinstance(arg, irast.CallArg):
+                self.mapped_args[node.path_id] = arg.expr.path_id
+                return setgen.ensure_set(
+                    irast.InlinedParameterExpr(
+                        typeref=arg.expr.typeref,
+                        required=node.expr.required,
+                        is_global=node.expr.is_global,
+                    ),
+                    path_id=arg.expr.path_id,
+                    ctx=self.ctx,
+                )
+            else:
+                return arg
```

Comment:
Default values get added to `inlined_args` as a `irast.Set` instead of `CallArg`. In that case, the default value is inlined directly. This should work because defaults are always constants.